addons/mail/models/mail_channel.py

```diff
@@ -149,25 +267,45 @@ def unlink(self):
 
     @api.multi
     def write(self, vals):
+        # First checks if user tries to modify moderation fields and has not the right to do it.
+        # We don't check if the fields touched are actually modified.
+        moderation_keys_are_touched = any(key for key in moderation_keys if vals.get(key))
+        if moderation_keys_are_touched and not self.env.user.has_group('base.group_system'):
+            if any(self.env.user not in channel.moderator_ids for channel in self if channel.moderation):
+                raise UserError("You do not possess the rights to modify fields related to moderation on one of the channels you are trying to modify!")
+
         tools.image_resize_images(vals)
         result = super(Channel, self).write(vals)
+
         if vals.get('group_ids'):
             self._subscribe_users()
+
+        if vals.get('moderation') is False:
+            MessagesToAccept = self.env['mail.message'].search(
+                        [
+                            ('moderation_status', '=', 'pending_moderation'),
+                            ('model', '=', 'mail.channel'),
+                            ('res_id', 'in', self.ids)
+                        ]
+                )
+            MessagesToAccept._accept_message()
+
         return result
 
     def get_alias_model_name(self, vals):
         return vals.get('alias_model', 'mail.channel')
 
     def _subscribe_users(self):
         for mail_channel in self:
-            mail_channel.write({'channel_partner_ids': [(4, pid) for pid in mail_channel.mapped('group_ids').mapped('users').mapped('partner_id').ids]})
+            partner_ids = mail_channel.mapped('group_ids').mapped('users').mapped('partner_id').ids
+            mail_channel.write({'channel_partner_ids': [(4, pid) for pid in partner_ids]})
 
     @api.multi
     def action_follow(self):
         self.ensure_one()
         channel_partner = self.mapped('channel_last_seen_partner_ids').filtered(lambda cp: cp.partner_id == self.env.user.partner_id)
         if not channel_partner:
-            return self.write({'channel_last_seen_partner_ids': [(0, 0, {'partner_id': self.env.user.partner_id.id})]})
+            return self.write({'channel_last_seen_partner_ids': [(0, _, {'partner_id': self.env.user.partner_id.id})]})
```

Comment:
Why this change ? Normally it does not change anythin, the second argument of a (0, blah, foo) is simply ignored.

Reply:
done